README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# baRcodeR

<!-- badges: start -->

[![Status](https://www.r-pkg.org/badges/version/baRcodeR)](https://CRAN.R-project.org/package=baRcodeR)
[![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/baRcodeR)](https://CRAN.R-project.org/package=baRcodeR)
[![Travis build
status](https://travis-ci.org/ropensci/baRcodeR.svg?branch=master)](https://travis-ci.org/ropensci/baRcodeR)
[![Codecov test
coverage](https://codecov.io/gh/ropensci/baRcodeR/branch/master/graph/badge.svg)](https://app.codecov.io/gh/ropensci/baRcodeR?branch=master)
[![Project Status: Active â€“ The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN
checks](https://cranchecks.info/badges/worst/baRcodeR)](https://cran.r-project.org/web/checks/check_results_baRcodeR.html)
[![rOpenSci
peer-review](https://badges.ropensci.org/338_status.svg)](https://github.com/ropensci/software-review/issues/338)
<!-- badges: end -->

baRcodeR generates labels for more repeatable workflows with biological
samples

## Installation

You can install the released version of baRcodeR from
[CRAN](https://CRAN.R-project.org) with:

    install.packages("baRcodeR")

And the development version from [GitHub](https://github.com/) with:

    # install.packages("devtools")
    devtools::install_github("ropensci/baRcodeR", build_vignettes = T)
    # for windows users to build vignettes
    # install_github("ropensci/baRcodeR", build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE)

> NOTE: Restarting RStudio is necessary for the addin for baRcodeR to
> appear.

## Quick Start

Text identifiers can be created in a sequential or hierarchical pattern.

``` r
library(baRcodeR)
```

    ## Loading required package: qrcode

``` r
example_labels <- uniqID_maker(user = FALSE, string = "Example", level = 1:80)
head(example_labels)
```

    ##        label ind_string ind_number
    ## 1 Example001    Example        001
    ## 2 Example002    Example        002
    ## 3 Example003    Example        003
    ## 4 Example004    Example        004
    ## 5 Example005    Example        005
    ## 6 Example006    Example        006

Then the text identifiers can be printed out with a laser printer on
sticker sheets.

``` r
pdf_file_name <- tempfile()
create_PDF(Labels = example_labels, name = pdf_file_name)
```

![](man/figures/example.png)<!-- -->

Th particular layout above defaults to ULINE 1.75” \* 0.5” labels but
other layouts can be specified through parameters in the
`custom_create_PDF` function.

## Introduction

`baRcodeR` is a R package for generating unique identifier strings and
printable 2D (QR) barcodes, with the aim of improving repeatability of
labelling, tracking and curating data from biological samples.
Specifically, users can:

-   generate simple ID codes (Ex001, Ex002, Ex003 …),
-   generate hierarchical (i.e. nested) ID codes (A01-B01, A01-B02,
    A02-B01, A02-B02, A03-B01 …),
-   generate printable PDF files of paired ID codes and QR barcodes with
    default spacing for ULINE 1.75” \* 0.5” WEATHER RESISTANT LABEL for
    laser printer; item # S-19297 (uline.ca)
-   customize the PDF layout for any type of printable format (e.g,
    vinyl stickers, waterproof paper)
-   generate reproducible code for archival purposes (e.g. in
    publications or online repositories)
-   create CSV files to link unique IDs and sampling hierarchy with
    downstream data collection workflows. For example, the PyTrackDat
    pipeline can be used to set up a web-based data collection platform:
    <https://github.com/pytrackdat/pytrackdat>

Creating unique, scannable barcodes generally involves two steps:

1.  Generate unique ID codes with `uniqID_maker()` or
    `uniqID_hier_maker()`
2.  Create a PDF file containing unique ID codes coupled with 2D barcode
    using `create_PDF()`

If you already have ID codes saved in a CSV file, the csv can be read
into a `data.frame()` in R. The `label` column, if it exists will be
used as input to generate barcodes. Otherwise, the first column in the
data frame will be used.

> NOTE: When printing from pdf, ensure that ‘anti-aliasing’ or
> ‘smoothing’ options are turned OFF, and that you are not using ‘fit to
> page’ or similar options that will re-scale the output.

![Flowchart of major functions](man/figures/Flowchart.png)

### Cheat Sheet

A 2-page, quick-reference guide is available via
[Figshare](https://dx.doi.org/10.6084/m9.figshare.7043309)

## Usage with RStudio addin

Please load the vignette “Use Addin”.

``` r
library(baRcodeR)
vignette("use-addin")
```

## Usage from the console

Please load the vignette “Using-baRcodeR” for console use.

``` r
vignette("Using-baRcodeR")
```

# Contribution

Please note that the ‘baRcodeR’ project is released with a [Contributor
Code of
Conduct](https://github.com/ropensci/baRcodeR/blob/master/CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.

Please document issues with a description, a minimal reproducible
example, and the `sessionInfo()`.

``` r
sessionInfo()
```

    ## R version 4.1.3 (2022-03-10)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 22000)
    ## 
    ## Matrix products: default
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_Canada.1252  LC_CTYPE=English_Canada.1252   
    ## [3] LC_MONETARY=English_Canada.1252 LC_NUMERIC=C                   
    ## [5] LC_TIME=English_Canada.1252    
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ## [1] baRcodeR_0.1.7 qrcode_0.1.4  
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] png_0.1-7         assertthat_0.2.1  digest_0.6.29     R.methodsS3_1.8.1
    ##  [5] magrittr_2.0.2    evaluate_0.15     highr_0.9         rlang_1.0.2      
    ##  [9] stringi_1.7.6     cli_3.2.0         rstudioapi_0.13   R.oo_1.24.0      
    ## [13] R.utils_2.11.0    rmarkdown_2.13    tools_4.1.3       stringr_1.4.0    
    ## [17] xfun_0.30         yaml_2.3.5        fastmap_1.1.0     compiler_4.1.3   
    ## [21] htmltools_0.5.2   knitr_1.38

# See also:

-   [zintr](https://github.com/carlganz/zintr)is an R interface to the C
    zint library. Use zintr if you want to create single barcode images.
    zintr does not include functions for (i) automating the creation of
    biologically-relevant, unique ID codes or (ii) customizable layouts
    for printing multiple barcodes.

-   [zint](https://zint.org.uk/) is a C library that generates a variety
    of different barcodes. Just like zintr, zint produces single barcode
    images.

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
# baRcodeR 0.1.7 

Thanks to rwalke for adding "Linear2" to the option list (Linear, Matrix). This will produce Code 128 extended linear barcodes, allowing for particular 1-D barcode patterns that are distinct from "Linear"

Thanks to Theirry() for pointing out that we need to change qrcode::qr_code() from (deprecated) qrcode_gen


# baRcodeR 0.1.6

At the request of a COVID-2019 research group, we have added an option to allow non-encoded text to appear with linear & 2D barcodes. These have been added to custom_create_PDF():

* alt_text -- adds human-readable text that is NOT encoded in the digital barcode
* denote -- characters used to denote non-encoded text

These are considered advanced features that should be used cautiously, and therefore they are not made available through the Addins GUI. 

# baRcodeR 0.1.5

Bugs:

- fixed errors in tests due to r-devel switching to using stringsAsFactor = FALSE as default
- fixed broken link in readme


# baRcodeR 0.1.4

Bugs and Improvements:

- tests have been added for user prompts and the RStudio addin with shinytest
- vignettes describing how to use the add-in and the command line are now organized separately from the restructured README
- `custom_create_PDF()` page generation should now be faster (helpful when making sheets for hundreds of labels)
- the command prompts have been restructured for a more menu-like selection for yes/no questions.
- other minor changes in addition to the major ones outlined above as suggested by reviews at rOpenSci documented [here](https://github.com/ropensci/software-review/issues/338)
- the baRcodeR package is now a part of [rOpenSci](https://ropensci.org/) and the documentation is online (here)[https://docs.ropensci.org/baRcodeR/]

# baRcodeR 0.1.3

Bugs and Improvements:

- major bug fix for linear barcodes that occasionally created unscannable barcodes.
- added documentation on how to create alternative formatting of labels (e.g. spaces, line breaks)
- added padding for labels which were single character or blank
- 2-page cheatsheet now available as addin 

# baRcodeR 0.1.2

New Feature:

- In response to a user request, there is now an option to print linear (code 128 set B) barcodes. 

# baRcodeR 0.1.1

Bugs and Improvements: 

- create_PDF() function will replace all underscores in text with dashes. Underscores are not specified in the encoding dictionary of `qcrode` and will throw errors.
- x_space and y_space parameters are now limited between 0 and 1 for easier use. These parameters are used to position text on the printed labels.
- Font size is no longer limited and is now measured as points. Font size is automatically reduced if text code is too long for the printed labels.

New Features:

- label_width and label_height parameters specify the width and height of the label to enable alleys (i.e. gaps) between physical labels.


-----------------


# baRcodeR 0.1.0

This is the first official release of the package.

tests/testthat.R
library(testthat)
library(baRcodeR)

test_check("baRcodeR")


tests/testthat/test_warnings.R
library(baRcodeR)

context("Input warnings")

test_that("input errors are produced", {
  expect_error(
    uniqID_maker(string="example", level="hi"), 
    "Level is not a string of numbers")
  
  expect_error(
    uniqID_maker(string="example", level = c(1:5), digits="a"), 
    "Digits is not a numerical value")
  
  expect_warning(
    uniqID_maker(string="example", level=1:300, digits=2), 
    "Digits specified less than max number. Increasing number of digits.")
  
  expect_warning(
    uniqID_maker(string="example", level=9:10, digits = 1), 
    "Digits specified less than max number. Increasing number of digits.")
 
  expect_error(
    uniqID_hier_maker(hierarchy = c("a",3,6)),
    "Hierarchy is not in list format")
  
   expect_error(
     uniqID_hier_maker(hierarchy = list(c("a",3,6), c("b",1))), 
     "Hierarchy entries are not of equal length.")
   
  expect_error(
    uniqID_hier_maker(hierarchy = list(c(3,6), c(1, 3))), 
    "Each level in hierarchy should have a string, a beginning value and an end value")
  
  expect_error(
    uniqID_hier_maker(hierarchy = list(c("a",3,6), c("b",1,"j"))), 
    "Invalid ending number on level")
  
  expect_error(
    uniqID_hier_maker(hierarchy = list(c("a",3,6), c("b","j",3))), 
    "Invalid starting number")
  
  expect_error(
    uniqID_hier_maker(hierarchy = list(c("a",3,6))), 
    "Input list has only one level")
  
  expect_warning(
    uniqID_hier_maker(hierarchy = list(c("a",3,6), c("b",100,103)), digits = 2),
    "Digits specified less than max level number")
  
  expect_warning(
    qrcode_make(1, "H"), 
    "Label is single character or blank.")
  
  temp_file <- tempfile()
  expect_error(
    custom_create_PDF(Labels = data.frame(label = c("ao1", "a02", "a03")), 
                      name = temp_file, 
                      numrow = "20"),
    "One or more numerical parameters are not numeric")
  expect_error(
    custom_create_PDF(
      Labels = data.frame(label = c("ao1", "a02", "a03")), 
      name = temp_file, 
      x_space = 1.5),
    "x_space value out of bounds. Must be between 0 and 1"
  )
  expect_error(
    custom_create_PDF(
      Labels = data.frame(label = c("ao1", "a02", "a03")), 
      name = temp_file, 
      y_space = 1.5),
    "y_space value out of bounds. Must be between 0 and 1"
  )

})



tests/testthat/test-test-addin.R
context("add-in tests")

library(shinytest)

test_that("labels", {
  skip_on_cran()
  expect_pass(testApp("apps/make_labels/", compareImages = FALSE))
})



R/zzz.R
.onAttach <- function(libname, pkgname){
  options(baRcodeR.connection = stdin())
}

fake_menu <- function(choices, prompt_string){
  attempt <- 1
  attempts_allowed <- 3
  # real menu doesn't work with std in
  choice_string <- paste(1:length(choices), ":", choices, collapse = "\n")
  repeat {
    choice <- numeric_input(paste0(prompt_string, "\n", choice_string))
    if (choice == 0 | choice > length(choices)){
      cat(paste0("Invalid input. Integer must be between 1 and ", length(choices)))
    }
    attempt <- attempt + 1
    if(attempt > attempts_allowed){
      stop("Invalid input. Please try again.")
    }
    if (choice <= length(choices) & choice != 0){
     break 
    }
  }
  return(choice)
}


numeric_input <- function(prompt_string, attempts_allowed = 3, integer = TRUE){
  attempt <- 1
  repeat {
    cat(prompt_string)
    startNum <- suppressWarnings(abs(as.numeric(readLines(con = getOption("baRcodeR.connection"), n = 1))))
    if(integer){
      startNum <- as.integer(startNum)
    }
    attempt <- attempt + 1
    # exit if input is number
    if (!is.na(startNum)) {
      break
    }
    # exit if too many attempts
    if(attempt > attempts_allowed){
      stop("Invalid input. Please try again.")
    } else {
      cat("Invalid input. Please enter a number")
    }
  }
  ifelse(!is.na(startNum), return(startNum))
  
}


string_input <- function(prompt_string){
  cat(prompt_string)
  return(readLines(con = getOption("baRcodeR.connection"), n = 1))
}


R/uniqIDMaker.R
#' Generate a list of ID codes
#'
#' Create ID codes consisting of a text string and unique numbers (string001, string002, ...). 
#' Can be run in interactive mode, prompting user for input. The data.frame 
#' output can be saved as CSV for (i) the \code{\link{create_PDF}} function 
#' to generate printable QR-coded labels; and (ii) to downstream data 
#' collection software (spreadsheets, relational databases, etc.)
#'
#' When the function is called with \code{user = TRUE}, a sequence of 
#' numbers is generated between the starting and ending number provided by the 
#' user. When \code{user = FALSE}, a vector of custom numbers can be provided. 
#' See example below.
#'
#' @return data.frame with text labels in the first column, along with string
#' and numeric values in two additional columns. 
#' @param user logical. Run function using interactive mode (prompts user for 
#' parameter values). Default is \code{FALSE}
#' @param string character. Text string for label. Default \code{null}.
#' @param level integer vector. Defines the numerical values to be appended
#' to the character string. Can be any sequence of numbers (see examples).
#' @param digits numerical. Default is \code{2}. Number of digits to be printed, 
#' adding leading 0s as needed. This will apply to all levels when \code{user=FALSE}. 
#' When the numeric value of the label has a greater number of digits than 
#' \code{digits}, \code{digits} is automatically increased for the entire level. 
#' Default is \code{3}.
#' @param ending_string a character string or vector of strings to attach to the label.
#' If a vector is used, all combinations of that vector with a unique label will be produced.
#'
#' @seealso \code{\link{uniqID_hier_maker}}
#' @export
#' @examples
#'
#'
#' ## sequential string of numbers in label
#' Labels <- uniqID_maker(string = "string", level = c(1:5), digits = 2)
#' Labels
#' 
#' ## can also use nonsequential strings in input for levels
#' level <- c(1:5, 8:10, 999:1000)
#' Labels <- uniqID_maker(string = "string", level = level, digits = 4)
#' Labels
#' 
#' 
#' ## Using the ending_string to produce labels with unique endings
#' ## this is different from hierarchical labels with two levels as there 
#' ## is no numbering, just the text string
#' 
#' 
#' Labels <- uniqID_maker(string = "string", level = c(1:5), digits = 2, ending_string = "A")
#' Labels
#' 
#' Labels <- uniqID_maker(string = "string", level = c(1:5), 
#'                        digits = 2, ending_string = c("A", "B"))
#' Labels
#' 
#'
#' if(interactive()){
#' ## function using user prompt does not use any of the other parameters
#' Labels <- uniqID_maker(user = TRUE)
#' Labels
#' }



uniqID_maker <- function(user = FALSE, string = NULL, level, digits = 3, ending_string = NULL){
  if (user == TRUE) {
    ## asks for string
    string <- string_input("Please enter string for level: ")
    ## first value of level
    startNum <- numeric_input("Please enter the starting number for level (integer): ")
    ## ending value of level
    endNum <- numeric_input("Enter the ending number for level: ")
    ## finds greatest value
    maxNum <- max(startNum,endNum)
    ## checks what is number of digits in max value
    digitsMax <- nchar(paste(maxNum))
    ## asks for how many digits to print
    digits <- numeric_input("Number of digits to print for level: ")
    level <-c(startNum:endNum)
    
  } 
  if (is.numeric(level) == FALSE) stop("Level is not a string of numbers")
  if (is.numeric(digits) == FALSE) stop("Digits is not a numerical value")
  ## if the number to digits to print is less than the digits in max value, increase digits
  if (nchar(paste(max(level))) > digits){
    warning("Digits specified less than max number. Increasing number of digits.")
    digits <- nchar(paste(max(level)))
  }
  
  if(!is.null(ending_string)){
    line <- paste0(string, "%0", digits, "d")
    label <- sprintf(line, rep(level))
    ind_string <- rep(string, length(rep(level)))
    ind_number <- sprintf(paste0("%0", digits, "d"), rep(level))
    suffix_strings <- expand.grid(label, ending_string)$Var2
    label <- paste(label, suffix_strings, sep = "-")
    label_df <- data.frame(label, ind_string, ind_number, end_string = as.character(suffix_strings), stringsAsFactors = FALSE)
  } else {
    line <- paste0(string, "%0", digits, "d")
    label <- sprintf(line, rep(level))
    ind_string <- rep(string, length(rep(level)))
    ind_number <- sprintf(paste0("%0", digits, "d"), rep(level))
    label_df <- data.frame(label, ind_string, ind_number, stringsAsFactors = FALSE)
  }
  
  
  return(label_df)
  
}






R/uniqIDHierarchy.R
#' Make hierarchical ID codes
#'
#' Generate hierarchical ID codes for barcode labels. 
#' Hierarchical codes have a nested structure: e.g. Y subsamples from 
#' each of X individuals. Use \code{\link{uniqID_maker}} 
#' for sequential single-level labels. Can be run in interactive mode, 
#' prompting user for input. The data.frame can be saved as CSV for 
#' (i) the \code{\link{create_PDF}} function to generate printable 
#' QR-coded labels; and (ii) to downstream data collection using spreadsheet, 
#' relational database, etc.
#'
#' @param user logical. Run function using interactive mode (prompts user for 
#' parameter values). Default is \code{FALSE}
#' @param hierarchy list. A list with each element consisting of three members
#'  a vector of three elements (string, beginning value, end value). See examples.
#'  Used only when \code{user=FALSE})
#' @param end character. A string to be appended to the end of each label.
#' @param digits numerical. Default is \code{2}. Number of digits to be printed, 
#' adding leading 0s as needed. This will apply to all levels when \code{user=FALSE}. 
#' When the max number of digits in the ID code is greater than number of digits 
#' defined in \code{digits}, then \code{digits} is automatically increased 
#' to avoid errors. 
#' @export
#' @return data.frame of text labels in the first column, with additional columns 
#' for each level in the hierarchy list, as defined by the user.
#' 
#' @seealso \code{\link{uniqID_maker}}
#' @examples
#' if(interactive()){
#' ## for interactive mode
#' uniqID_hier_maker(user = TRUE)
#' }
#'
#' ## how to make hierarchy list
#'
#' ## create vectors for each level in the order string_prefix, beginning_value,
#' ## end_value and combine in list
#'
#' a <- c("a", 3, 6)
#' b <- c("b", 1, 3)
#' c <- list(a, b)
#' Labels <- uniqID_hier_maker(hierarchy = c)
#' Labels
#'
#' ## add string at end of each label
#' Labels <- uniqID_hier_maker(hierarchy = c, end = "end")
#' Labels
#'


uniqID_hier_maker <- function(user = FALSE, hierarchy, end = NULL, digits = 2){
  # user interaction code
  if(user == TRUE){
    hlevels <- numeric_input("What is the number of levels in hierarchy: ")
    
    strEnd <- switch(fake_menu(c("Yes", "No"), "String at end of label? "), TRUE, FALSE)
    
    if(strEnd){
      end <- string_input("Please enter ending string: ")
    } else {
      end <- ""
    }
    
    digits <- numeric_input("Number of digits to print: ")
    
    # possible inputs

    hierarchy <- vector("list", hlevels)
    
    for(i in seq(1,hlevels)){
      str <- string_input(paste0("Please enter string for level ",i,": "))
      
      startNum <- numeric_input(paste0("Enter the starting number for level ",i,": "))
      
      endNum <- numeric_input(paste0("Enter the ending number for level ",i,": "))
      
      maxNum <- max(startNum,endNum)
      hierarchy[[i]]<-c(str, startNum, endNum)
    }
  }
  # end user input
  # hierarchy format check
  if (is.list(hierarchy) == FALSE) stop("Hierarchy is not in list format. See ?uniqID_hier_maker")
  if (length(unique(vapply(hierarchy, length, integer(1)))) != 1) stop("Hierarchy entries are not of equal length.")
  if (any(vapply(hierarchy, length, integer(1)) != 3)) 
    stop("Each level in hierarchy should have a string, a beginning value and an end value.")
  if (length(hierarchy) == 1) 
    stop("Input list has only one level. Did you forget a level or are you sure you are not looking for uniqIDMaker()?")
  # loop through hierarchy to generate text
  for(i in seq(1,length(hierarchy))){
    str <- hierarchy[[i]][1]
    startNum <- suppressWarnings(as.numeric(hierarchy[[i]][2]))
    endNum <- suppressWarnings(as.numeric(hierarchy[[i]][3]))
    if (is.na(startNum) == TRUE){
      stop(paste0("Invalid starting number on level ", i, ". Please doublecheck your input"))
    }
    if (is.na(endNum) == TRUE){
      stop(paste0("Invalid ending number on level ", i, ". Please doublecheck your input"))
    }
    maxNum <- max(startNum, endNum)
    digitsMax <- max(digits, nchar(paste(maxNum)))
    if (digitsMax > digits){
      warning("Digits specified less than max level number. Increasing number of digits for level")
      digits <- digitsMax
    }
    lvlRange <- c(startNum:endNum)
    line <- paste0(str, "%0", digits, "d")
    Labels <- sprintf(line, rep(lvlRange))
    if (i > 1) {
      ## replicate previous labels by number of elements in present level
      temp_barcode <- rep(barcodes, each = length(startNum:endNum))
      ## rep present level elements so the length matches that of temp_barcode
      temp_label <- rep(Labels, length.out = length(temp_barcode))
      ## paste everything together
      Labels <- paste(temp_barcode, temp_label, sep = "-")

    }
    barcodes <- Labels
  } # end hierarchy making loop
  # makes columns out of hierarchy levels
  label_df <- cbind(barcodes, data.frame(t(sapply(strsplit(barcodes, "-"),c)), stringsAsFactors = FALSE))
  df_names <- sapply(hierarchy, function(x) x[1])
  names(label_df) <- c("label", df_names)
  # dont forget to add the string at the end
  label_df$label <- paste0(label_df$label, end)
  return(label_df)
}



R/createPDF.R
#' Make barcodes and print labels
#'
#' Input vector or data.frame of ID codes to produce a PDF of QR codes which can
#' be printed. This is a wrapper function for \code{\link{custom_create_PDF}}.
#' See details of \code{\link{custom_create_PDF}} on how to format text labels
#' if needed.
#' 
#' The default PDF setup is for ULINE 1.75" * 0.5" WEATHER RESISTANT LABEL for laser
#' printer; item # S-19297 (uline.ca). The page format can be modified using
#' the \code{...} (advanced arguments) for other label types.
#'
#' @return a PDF file containing QR-coded labels, saved to the default directory.
#'
#'
#' @inheritParams custom_create_PDF
#' @param ... advanced arguments to modify the PDF layout. See
#'  \code{\link{custom_create_PDF}} for arguments. The advanced options can be
#'   accessed interactively with \code{user = TRUE} and then entering TRUE when prompted to
#'    modify advanced options.
#' @export
#' @examples
#' ## data frame
#' example_vector <- as.data.frame(c("ao1", "a02", "a03"))
#'
#' \dontrun{
#' ## run with default options
#' ## pdf file will be "example.pdf" saved into a temp directory
#' 
#' temp_file <- tempfile()
#' 
#' create_PDF(Labels = example_vector, name = temp_file)
#' 
#' ## view example output from temp folder
#' system2("open", paste0(temp_file, ".pdf"))
#' }
#' 
#' ## run interactively. Overrides default pdf options
#' if(interactive()){
#'     create_PDF(user = TRUE, Labels = example_vector)
#' }
#' 
#' \dontrun{
#' ## run using a data frame, automatically choosing the "label" column
#' example_df <- data.frame("level1" = c("a1", "a2"), "label" = c("a1-b1",
#' "a1-b2"), "level2" = c("b1", "b1"))
#' create_PDF(user = FALSE, Labels = example_df, name = file.path(tempdir(), "example_2"))
#' }
#' 
#' \dontrun{
#' ## run using an unnamed data frame
#' example_df <- data.frame(c("a1", "a2"), c("a1-b1", "a1-b2"), c("b1", "b1"))
#' ## specify column from data frame
#' create_PDF(user = FALSE, Labels = example_df[,2], name = file.path(tempdir(), "example_3"))
#' }
#' \dontrun{
#' ## create linear (code128) label rather than matrix (2D/QR) labels
#' example_df <- data.frame(c("a1", "a2"), c("a1-b1", "a1-b2"), c("b1", "b1"))
#' ## specify column from data frame
#' create_PDF(user = FALSE, Labels = example_df, name = file.path(tempdir(),
#' "example_4", type = "linear"))
#' }
#' @seealso \code{\link{custom_create_PDF}}


create_PDF <- function(user = FALSE,
                     Labels = NULL,
                     name ="LabelsOut",
                     type = "matrix",
                     ErrCorr = "H",
                     Fsz = 12, ...) {
  custom_create_PDF(user, Labels, name, type, ErrCorr, Fsz, ...)
}




